syntax = "proto3";

package secrets_manager.v1;

import "google/protobuf/timestamp.proto";

// Secrets Manager gRPC API
service SecretsManagerService {
    // Secret CRUD
    // Create a new secret entry.
    rpc CreateSecret(CreateSecretRequest) returns (Secret);

    // Get an existing secret by key_name.
    // Returns decrypted value.
    rpc GetSecretByName(GetSecretByNameRequest) returns (Secret);

    // Asymmetric Key Pair CRUD
    // Create a new asymmetric key pair (public/private).
    rpc CreateAsymmetricKeyPair(CreateAsymmetricKeyPairRequest) returns (AsymmetricKeyPair);

    // Retrieve an asymmetric key pair by key_name.
    // Optionally return only public or both keys.
    rpc GetAsymmetricKeyPairByKeyName(GetAsymmetricKeyPairByKeyNameRequest) returns (AsymmetricKeyPair);

    // Symmetric Key CRUD
    // Create a new symmetric key.
    rpc CreateSymmetricKey(CreateSymmetricKeyRequest) returns (SymmetricKey);

    // Retrieve a symmetric key by key_name.
    rpc GetSymmetricKeyByKeyName(GetSymmetricKeyByKeyNameRequest) returns (SymmetricKey);
}

// Response

message Secret {
  string kid = 1;
  string key_name = 2;
  string key_type = 3;
  // Plaintext secret value as UTF-8 string.
  // (Internally encrypted and stored as bytes in DB.)
  string value = 4;
  google.protobuf.Timestamp created_at = 5;
  int32 mk_version = 6;
  bool is_active = 7;
}

message AsymmetricKeyPair {
  string kid = 1;
  string key_name = 2;
  string key_type = 3;
  // PEM-encoded key material as UTF-8 string.
  string public_key = 4;
  // PEM-encoded key material as UTF-8 string. Empty when return_key_type="public".
  string private_key = 5;
  google.protobuf.Timestamp created_at = 6;
  int32 mk_version = 7;
  bool is_active = 8;
}

message SymmetricKey {
  string kid = 1;
  string key_name = 2;
  string key_type = 3;
  // Base64-encoded raw key bytes.
  string key_value = 4;
  google.protobuf.Timestamp created_at = 5;
  int32 mk_version = 6;
  bool is_active = 7;
}


// Requests


message CreateSecretRequest {
  string key_name = 1;
  string value = 2;

  // Optional metadata
  string key_type = 3;
  int32 mk_version = 4;
}

message CreateAsymmetricKeyPairRequest {
  int32 key_size = 1;
  string key_name = 2;

  // Optional metadata
  string key_type = 3;
  int32 mk_version = 4;
}

message CreateSymmetricKeyRequest {
  int32 key_size = 1;
  string key_name = 2;

  // Optional metadata
  string key_type = 3;
  int32 mk_version = 4;
}


message GetSecretByNameRequest {
  string key_name = 1;
}

message GetAsymmetricKeyPairByKeyNameRequest {
  string key_name = 1;

  // "public" | "pair" (matches service.py default: "public")
  string return_key_type = 2;
}

message GetSymmetricKeyByKeyNameRequest {
  string key_name = 1;
}
